src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.h

```diff
@@ -0,0 +1,264 @@
+#include "crimson/common/errorator.h"
+#include "crimson/os/seastore/extentmap_manager.h"
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h"
+
+namespace crimson::os::seastore::extentmap_manager {
+
+/**
+ * ExtMapInnerlNode
+ *
+ * Abstracts operations on and layout of internal nodes for the
+ * Extentmap Tree.
+ *
+ * Layout (4k):
+ *   num_entries: uint16_t           2b
+ *   (padding)  :                    14b
+ *   keys       : objaddr_t[340]     (340*4)b
+ *   values     : laddr_t[340]       (340*8)b
+ *                                    = 4096
+ */
+constexpr size_t INNER_NODE_CAPACITY = 340;
+
+struct ExtMapInnerNode
+  : ExtMapNode,
+    common::FixedKVNodeLayout<
+      INNER_NODE_CAPACITY,
+      objaddr_t, ceph_le32,
+      laddr_t, laddr_le_t> {
+  using internal_iterator_t = const_iterator;
+  template <typename... T>
+  ExtMapInnerNode(T&&... t) :
+    ExtMapNode(std::forward<T>(t)...),
+    FixedKVNodeLayout(get_bptr().c_str()) {}
+
+  static constexpr extent_types_t type = extent_types_t::EXTMAP_INNER;
+
+  CachedExtentRef duplicate_for_write() final {
+    assert(delta_buffer.empty());
+    return CachedExtentRef(new ExtMapInnerNode(*this));
+ };
+
+  delta_buffer_t delta_buffer;
+  delta_buffer_t *maybe_get_delta_buffer() {
+    return is_mutation_pending() ? &delta_buffer : nullptr;
+  }
+
+  seek_lextent_ret seek_lextent(Transaction &t, objaddr_t lo, extent_len_t len) final;
+
+  insert_ret insert(Transaction &transaction,
+                    objaddr_t lo, lext_map_val_t val) final;
+
+  rm_lextent_ret rm_lextent(Transaction &t, objaddr_t lo, lext_map_val_t val) final;
+
+  punch_lextent_ret punch_lextent(Transaction &t, objaddr_t lo, extent_len_t len) final;
+
+  find_hole_ret find_hole(Transaction &t, objaddr_t lo, extent_len_t len) final;
+
+  split_children_ret make_split_children(Transaction &t) final;
+
+  full_merge_ret
+  make_full_merge(Transaction &t, ExtMapNodeRef right) final;
+
+  make_balanced_ret
+  make_balanced(Transaction &t, ExtMapNodeRef _right, bool prefer_left) final;
+
+  std::ostream &print_detail(std::ostream &out) const final;
+
+  extent_types_t get_type() const final {
+    return type;
+  }
+
+  ceph::bufferlist get_delta() final {
+    assert(!delta_buffer.empty());
+    ceph::buffer::ptr bptr(delta_buffer.get_bytes());
+    delta_buffer.copy_out(bptr.c_str(), bptr.length());
+    ceph::bufferlist bl;
+    bl.push_back(bptr);
+    return bl;
+  }
+
+  void apply_delta(const ceph::bufferlist &_bl) final {
+    assert(_bl.length());
+    ceph::bufferlist bl = _bl;
+    bl.rebuild();
+    delta_buffer_t buffer;
+    buffer.copy_in(bl.front().c_str(), bl.front().length());
+    buffer.replay(*this);
+  }
+
+  bool at_max_capacity() const final {
+    return get_size() == get_capacity();
+  }
+
+  bool at_min_capacity() const {
+    return get_size() == get_capacity() / 2;
+  }
+
+// should make it common
```

Comment:
Aren't they actually different?

Reply:
I think it only input parameter type is different, extentmp tree is  uint32_t, lba tree is is uint64_t.

Reply:
The conditions are different reflecting the fact that *Leaf elements have lengths (IIRC).